### PR TITLE
[FEATURE] Enhance email registration hook in RegistrationManager

### PR DESCRIPTION
**Pull request by mk-mxp**
Converts the previously existing hook to the new HookProvider and adds some parameters to enhance usability.

I will add docs and CHANGELOG.md later to this PR.

### CODE REVIEW

**Comment by mk-mxp** [comment]
Please help with this:

In the tests I could not get this to work:

```php
        $hook = $this->createMock(RegistrationEmail::class);
```

Instead I had to use this approach:
```php
        $hookClassName = 'RegistrationEmailHook' . \uniqid('', false);
        $hook = $this->getMockBuilder(RegistrationEmail::class)
            ->setMockClassName($hookClassName)->getMock();
```

But I don't understand why. In the first 2 new tests the default `createMock()` worked fine, all mocked methods got called. When adding more tests the HookProvider still was called (I could provoke exceptions there) but not the registered mocks. Do You know any reason for that?

**Comment by oliverklee** [comment]
I had similar problems and still do not fully understand this. It might have something to do with `GeneralUtitlity::makeInstance`.

I propose creating a dedicated fixture class that implements the interface.

**Comment by mk-mxp** [comment]
A subtile case of [PEBKAC](http://ars.userfriendly.org/cartoons/?id=19980506)...

If in a test `GeneralUtitlity::makeInstance()` **does not** take the instance added by `GeneralUtitlity::addInstance` (be it in a failed test or intended), it remains in the instance buffer. And the next test using the same class or interface mock will get that instance instead of the one it added.

Calling `GeneralUtitlity::makeInstance()` unconditionally to clean up the instance FIFO buffer at the end of the test solves this for most cases. If there is an exception aborting the test, this will not help.

To have all cases covered, `tearDown()` would need to purge the instances from the FIFO buffer (but there is only a `GeneralUtitlity::purgeInstances()` removing all known instances making all tests fail).

**Comment by mk-mxp** [comment]
Ready for review.

File: Classes/Hooks/Interfaces/RegistrationEmail.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ *
+ * Customize emails before they are sent.
+ *
+ * @author Pascal Rinker <projects@jweiland.net>
+ * @author Michael Kramer <m.kramer@mxp.de>
+ */
+interface RegistrationEmail extends Hook
+{
+    /**
+     * Modifies the attendee "Thank You" email, just before it is sent.
```
Comment (oliverklee) [inline Classes/Hooks/Interfaces/RegistrationEmail.php]: "Thank You" -> "Thank you" (also in the other places).

Comment (oliverklee) [inline Classes/Hooks/Interfaces/RegistrationEmail.php]: Please drop this comma (also in the other similar occurrences).

File: Classes/Hooks/Interfaces/RegistrationEmail.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+     *
+     * You may modify receiver or sender as well as subject and body of the eMail.
```
Comment (oliverklee) [inline Classes/Hooks/Interfaces/RegistrationEmail.php]: eMail -> email (in all affected places)

File: Classes/Hooks/Interfaces/RegistrationEmail.php [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+     * You may modify or set marker values in the template.
+     *
+     * @param \Tx_Oelib_Template $emailTemplate
+     * @param \Tx_Seminars_OldModel_Registration $registration
+     * @param string $emailReason Possible values:
+     *          - confirmation
+     *          - confirmationOnUnregistration
+     *          - confirmationOnRegistrationForQueue
+     *          - confirmationOnQueueUpdate
+     * @param bool $useHtml whether to create HTML instead of plain text
```
Comment (oliverklee) [inline Classes/Hooks/Interfaces/RegistrationEmail.php]: For the new hooks, please let's have two separate methods: one for the plain-text version and one for the HTML version (as the behavior-switching boolean parameter is an antipattern).

File: Documentation/DE/Referenz/Hooks/Index.rst [hunk 4]
```diff
@@ -60,7 +60,7 @@ machen Sie seminars in :file:`ext_localconf.php` Ihrer Extension bekannt:
 
 .. code-block:: php
 
-    $GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars'][\OliverKlee\Seminars\Hooks\Interfaces\SeminarSingleView:class][]
+    $GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars'][\OliverKlee\Seminars\Hooks\Interfaces\SeminarSingleView::class][]
```
Comment (oliverklee) [inline Documentation/DE/Referenz/Hooks/Index.rst]: I've fixed the syntax errors in the docs in a separate PR #447. Could you please rebase?

File: Documentation/DE/Referenz/Hooks/Index.rst [hunk 5]
```diff
@@ -385,83 +385,134 @@ Implementieren Sie die benötigten Methoden gemäß dem Interface:
...
 .. _notificationemail_de:
 
-Hooks to post process notification emails
-""""""""""""""""""""""""""""""""""""""""""
+Hooks zu den Emails der Registrierungsbenachrichtigungen
+""""""""""""""""""""""""""""""""""""""""""""""""""""""""
 
-To use this hook, please create a class that implements the interface
-\\OliverKlee\\Seminars\\Hooks\\RegistrationEmailHookInterface. Then you need to add the following methods:
+Es gibt 4 Hooks in die Emails der Registrierungsbenachrichtigungen:
```
Comment (oliverklee) [inline Documentation/DE/Referenz/Hooks/Index.rst]: I'd say "für die" instead of "in die" here. WDYT?

File: Documentation/DE/Referenz/Hooks/Index.rst [hunk 6]
```diff
@@ -385,83 +385,134 @@ Implementieren Sie die benötigten Methoden gemäß dem Interface:
...
 
-**Hook to post process the attendee email**
+* Bevor das Template für die Teilnehmer-Benachrichtigung in HTML oder Plain-Text umgewandelt wird
+* Bevor die Teilnehmer-Benachrichtigung abgeschickt wird
+* Bevor die Benachrichtigung an die Organisatoren abgeschickt wird
+* Bevor zusätzliche Statusinformationen an die Organisatoren abgeschickt werden
```
Comment (oliverklee) [inline Documentation/DE/Referenz/Hooks/Index.rst]: Bevor -> bevor (as these are no complete sentences)

File: Documentation/DE/Referenz/Hooks/Index.rst [hunk 7]
```diff
@@ -385,83 +385,134 @@ Implementieren Sie die benötigten Methoden gemäß dem Interface:
...
 
-.. code-block:: php
+Im Template-Hook können Sie eigene Marker ausfüllen oder vorhandene Marker-Werte ändern. Zu
+verfügbaren Eigenschaften und Methoden dafür siehe :file:`Classes/OldModel/Registration.php`.
+Dieser Hook wird doppelt aufgerufen, wenn HTML-Emails versandt werden. Der erster Aufruf ist für
+die HTML-Version, der Zweite für die Plain-Text-Variante.
```
Comment (oliverklee) [inline Documentation/DE/Referenz/Hooks/Index.rst]: This needs to be adapted once we have separate methods for this.

File: Tests/LegacyUnit/Service/RegistrationManagerTest.php [hunk 8]
```diff
@@ -2212,6 +2213,37 @@ public function notifyAttendeeForSendConfirmationFalseNotCallsModifyThankYouEmai
...
+        $GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars'][RegistrationEmail::class][] = $hookClass;
+        GeneralUtility::addInstance($hookClass, $hook);
+
+        $this->subject->setConfigurationValue('sendConfirmation', true);
+        $pi1 = new \Tx_Seminars_FrontEnd_DefaultController();
+        $pi1->init();
+
+        $this->subject->notifyAttendee($registration, $pi1);
+
+        GeneralUtility::makeInstance($hookClass);  // purge added instance in case of failing test
```
Comment (oliverklee) [inline Tests/LegacyUnit/Service/RegistrationManagerTest.php]: As an assertion failure exits the test method, this needs to go into `tearDown`.

Comment (oliverklee) [inline Tests/LegacyUnit/Service/RegistrationManagerTest.php]: No end-line comments, please -> please move the comment to the line above the code line.

File: Tests/LegacyUnit/Service/RegistrationManagerTest.php [hunk 9]
```diff
@@ -2237,6 +2269,39 @@ public function notifyAttendeeForSendConfirmationTrueAndPlainTextEmailCallsPostP
...
```
Comment (oliverklee) [inline Tests/LegacyUnit/Service/RegistrationManagerTest.php]: Ditto.

File: Tests/LegacyUnit/Service/RegistrationManagerTest.php (hunk 8 above)
Comment (mk-mxp) [reply Tests/LegacyUnit/Service/RegistrationManagerTest.php]: That is false as this does not use an `assert[...]` method. The assertion on mocked methods being called is evaluated **after** the test method has ended. Which includes the cleanup.

Moving this into `tearDown()` also raises the problem: What classname(s) were generated? They are random names. And creating a mock instance in `tearDown()` for each and every test method just to get the name? Running `::makeInstance()` that often for nothing? Or is there a way to attach a specialized `tearDown()` to a test I don't know about, yet?

File: Documentation/DE/Referenz/Hooks/Index.rst (hunk 5 above)
Comment (mk-mxp) [reply Documentation/DE/Referenz/Hooks/Index.rst]: It's like that in the other hook docs, too. It's an english word in a german sentence, and in English I would say "hook-in to" -> "einhaken in". But I don't really care. Less work is keep it.

File: Tests/LegacyUnit/Service/RegistrationManagerTest.php (hunk 8 above)
Comment (mk-mxp) [reply Tests/LegacyUnit/Service/RegistrationManagerTest.php]: Solved that.

File: Classes/Hooks/Interfaces/RegistrationEmail.php [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        \Tx_Oelib_Mail $mail,
+        \Tx_Seminars_Model_Registration $registration,
+        string $emailReason
+    );
+
+    /**
+     * Modifies the attendee "Thank you" email body just before the subpart is rendered to PlainText.
+     *
+     * This method is called for every confirmation mail, even if HTML emails are configured.
+     * The body of a HTML mail alwyas contains a PlainText version, too.
```
Comment (oliverklee) [inline Classes/Hooks/Interfaces/RegistrationEmail.php]: I propose we reword this to make it clear that this hook is called for the plain text part of all HTML emails as well.

File: Tests/LegacyUnit/Service/RegistrationManagerTest.php [hunk 11]
```diff
@@ -162,12 +168,19 @@ protected function setUp()
...
+
+        $this->mockedClassName = '';
     }
 
     protected function tearDown()
     {
         $this->testingFramework->cleanUp();
 
+        if ($this->mockedClassName !== '') {
+            GeneralUtility::makeInstance($this->mockedClassName);
```
Comment (oliverklee) [inline Tests/LegacyUnit/Service/RegistrationManagerTest.php]: I propose adding a comment why this needs to be done.

File: Tests/LegacyUnit/Service/RegistrationManagerTest.php [hunk 12]
```diff
@@ -2212,6 +2225,39 @@ public function notifyAttendeeForSendConfirmationFalseNotCallsModifyThankYouEmai
...
+        GeneralUtility::addInstance($hookClass, $hook);
+
+        $this->subject->setConfigurationValue('sendConfirmation', true);
+        $pi1 = new \Tx_Seminars_FrontEnd_DefaultController();
+        $pi1->init();
+
+        $this->subject->notifyAttendee($registration, $pi1);
+
+        // purge added instance in case of failing test
+        $this->mockedClassName = $hookClass;
```
Comment (oliverklee) [inline Tests/LegacyUnit/Service/RegistrationManagerTest.php]: For all tests, please let's move this directly to below the `addInstance` call so any exceptions etc. in the code do not break the cleanup.

File: Tests/LegacyUnit/Service/RegistrationManagerTest.php [hunk 13]
```diff
@@ -2212,6 +2225,39 @@ public function notifyAttendeeForSendConfirmationFalseNotCallsModifyThankYouEmai
...
+        $hook->expects(self::never())->method('modifyAttendeeEmailBodyHtml');
+        $hook->expects(self::never())->method('modifyOrganizerEmail');
+        $hook->expects(self::never())->method('modifyAdditionalEmail');
+
+        $hookClass = \get_class($hook);
+        $GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars'][RegistrationEmail::class][] = $hookClass;
+        GeneralUtility::addInstance($hookClass, $hook);
+
+        $this->subject->setConfigurationValue('sendConfirmation', true);
+        $pi1 = new \Tx_Seminars_FrontEnd_DefaultController();
```
Comment (oliverklee) [inline Tests/LegacyUnit/Service/RegistrationManagerTest.php]: `$pi1` -> `$controller` (also in the other new tests)

File: Tests/LegacyUnit/Service/RegistrationManagerTest.php [hunk 14]
```diff
@@ -2237,6 +2283,40 @@ public function notifyAttendeeForSendConfirmationTrueAndPlainTextEmailCallsPostP
...
+
+        $registration = $this->createRegistration();
+
+        $hook = $this->createMock(RegistrationEmail::class);
+        $hook->expects(self::once())->method('modifyAttendeeEmail')
+            ->with(self::anything(), self::anything(), 'confirmation');
+        $hook->expects(self::once())->method('modifyAttendeeEmailBodyPlainText')
+            ->with(self::anything(), $registration, 'confirmation');
+        $hook->expects(self::once())->method('modifyAttendeeEmailBodyHtml')
+            ->with(self::anything(), $registration, 'confirmation');
```
Comment (oliverklee) [inline Tests/LegacyUnit/Service/RegistrationManagerTest.php]: Please let's be a bit more specific here in the assertions. I think `self::instanceOf(…)` might work. And the registration instance might also work.

File: Tests/LegacyUnit/Service/RegistrationManagerTest.php (hunk 12 above)
Comment (mk-mxp) [reply Tests/LegacyUnit/Service/RegistrationManagerTest.php]: I refactored that to be more generic and as a direct drop-in replacemant for `GeneralUtility::makeInstance()`: `$this->addMockedInstance()`.

File: Tests/LegacyUnit/Service/RegistrationManagerTest.php [hunk 15]
```diff
@@ -169,17 +169,14 @@ protected function setUp()
 
         $this->frontEndUserMapper = \Tx_Oelib_MapperRegistry::get(\Tx_Seminars_Mapper_FrontEndUser::class);
 
-        $this->mockedClassName = '';
+        $this->mockedClassNames = [];
```
Comment (oliverklee) [inline Tests/LegacyUnit/Service/RegistrationManagerTest.php]: This line is not necessary is this already is the default value.

File: Tests/LegacyUnit/Service/RegistrationManagerTest.php [hunk 16]
```diff
@@ -321,6 +318,45 @@ private function filterEmailAttachmentsByTitle(MailMessage $email, string $conte
...
+     * Purges possibly leftover instances from the Typo3 instance FIFO buffer used by
+     * `GeneralUtility::makeInstance()`.
+     *
+     * @return void
+     */
+    private function purgeMockedInstances()
+    {
+        if (empty($this->mockedClassNames)) {
+            return;
+        }
```
Comment (oliverklee) [inline Tests/LegacyUnit/Service/RegistrationManagerTest.php]: We don't really need this guard clause here as a `foreach` on an empty array is a no-op. (This will simplify the code a bit.)

File: Tests/LegacyUnit/Service/RegistrationManagerTest.php [hunk 17]
```diff
@@ -2233,29 +2336,37 @@ public function notifyAttendeeForSendConfirmationTrueCallsRegistrationEmailHookM
...
 
         $hook = $this->createMock(RegistrationEmail::class);
-        $hook->expects(self::once())->method('modifyAttendeeEmail')
-            ->with(self::anything(), self::anything(), 'confirmation');
-        $hook->expects(self::once())->method('modifyAttendeeEmailBodyPlainText')
-            ->with(self::anything(), $registration, 'confirmation');
+        $hook->expects(self::once())->method('modifyAttendeeEmail')->with(
+            self::isInstanceOf(\Tx_Oelib_Mail::class),
+            self::identicalTo($registration),
+            self::identicalTo('confirmation')
```
Comment (oliverklee) [inline Tests/LegacyUnit/Service/RegistrationManagerTest.php]: AFAIK, `self::identicalTo(foo)` can be shortened to `foo`.
